hw/ip/otbn/dv/rig/rig/gens/known_wdr.py

```diff
@@ -0,0 +1,109 @@
+            raise RuntimeError('BN.NOT instruction not in instructions file')
+
+        self.insn_xor = insns_file.mnemonic_to_insn['bn.xor']
+        self.insn_not = insns_file.mnemonic_to_insn['bn.not']
+        # BN.XOR has six operands: wrd, wrs1, wrs2, shift_type, shift_value
+        # and flag_group
+        if not (len(self.insn_xor.operands) == 6 and
+                isinstance(self.insn_xor.operands[0].op_type, RegOperandType) and
+                self.insn_xor.operands[0].op_type.reg_type == 'wdr' and
+                self.insn_xor.operands[0].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[1].op_type, RegOperandType) and
+                self.insn_xor.operands[1].op_type.reg_type == 'wdr' and
+                not self.insn_xor.operands[1].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[2].op_type, RegOperandType) and
+                self.insn_xor.operands[2].op_type.reg_type == 'wdr' and
+                not self.insn_xor.operands[2].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[4].op_type, ImmOperandType)):
+            raise RuntimeError('BN.XOR instruction from instructions file is '
+                               'not the shape expected by the SmallVal generator.')
+
+        self.wrd_op_type = self.insn_xor.operands[0].op_type
+        self.wrs1_op_type = self.insn_xor.operands[1].op_type
+        self.imm_op_type = self.insn_xor.operands[4].op_type
+        self.imm_op_type.shift = 3
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this is the last instruction in the current gap
+        # because we need to either jump or do an ECALL to avoid getting stuck.
+        if program.get_insn_space_at(model.pc) <= 3:
+            return None
+
+        # The bn.xor-bn.not pair takes 2 instructions
+        if program.space < 2:
+            return None
+
+        if model.fuel < 2:
+            return None
+
+        # Pick wrd any old way: we can write to any register. This should
+        # always succeed.
+        wrd_val_xor = model.pick_reg_operand_value(self.wrd_op_type)
+        assert wrd_val_xor is not None
+
+        # Pick an operand value.
+        wrs1_val_xor = model.pick_reg_operand_value(self.wrs1_op_type)
+        assert wrs1_val_xor is not None
+
+        shift_bits = self.imm_op_type.op_val_to_enc_val(0, model.pc)
+        assert shift_bits is not None
+
+        shift_type = random.randint(0, 1)
+
+        wrd_val_not = model.pick_reg_operand_value(self.wrd_op_type)
+        assert wrd_val_not is not None
+
+        # Trying to create a program where bn.xor and bn.not are consequently
+        # executed
+
+        op_vals_xor = [wrd_val_xor, wrs1_val_xor, wrs1_val_xor, shift_type,
+                       shift_bits, 0]
+
+        op_vals_not = [wrd_val_not, wrd_val_xor, shift_type, shift_bits, 0]
```

Comment:
The zero at the end of each instruction's operands is a flag group argument. I'd probably pick that randomly (no real need to force it to zero). You can use essentially the same code as for shift_type. (As a bonus, that also lets us give a name to the value).